### ISSUE TITLE
Getting a lot less acknowledgements than I've sent messages

### ISSUE THREAD

**Issue by enchantner** [confluentinc/confluent-kafka-go#110]
Description
===========
I have a question pretty similar to https://github.com/confluentinc/confluent-kafka-go/issues/109 , but a little bit different. So, I use a Channel Based Producer, and I'm trying to use the same _select {}_ both to send messages and receive events. 

At the same time I'm counting both numbers - how many messages were sent to _ProduceChannel()_ and how many successful deliveries I got. After some time running, the second number is way (tens of thousands) behind the first one, and that's without a single error. After some more time the second number just stops updating at all, but first one still keeps increasing. At the end the whole code hangs, because I got far less acknoledgements than expected.

How to reproduce
================

`producer, err := kafka.NewProducer(
		&kafka.ConfigMap{
			"bootstrap.servers": strings.Join(conf.Conf.Kafka.Brokers, ", "),
			"message.max.bytes": 1024 * 1024 * 40,
			"go.batch.producer": true,
		})

go func() {
		defer close(producerDone)

		run := true
		for run {
			select {
				case <-cmdCh:
					run = false
				case chunk := <- resCh:
					producerCounter++
					if producerCounter % 10000 == 0 {
						logger.Info.Printf("Produced %d messages so far\n", producerCounter)
					}
					producer.ProduceChannel() <- &kafka.Message{
						TopicPartition: kafka.TopicPartition{
							Topic:     &conf.Conf.Kafka.Topic,
							Partition: kafka.PartitionAny},
						Value: chunk}
				case ev := <- producer.Events():
					switch ev := ev.(type) {
					case *kafka.Message:
						m := ev
						if m.TopicPartition.Error != nil {
							logger.Info.Printf("Delivery failed: %v\n", m.TopicPartition.Error)
						} else {
							ackCounter++
							if ackCounter % 10000 == 0 {
								logger.Info.Printf("ACKed %d messages so far\n", ackCounter)
							}
						}
					default:
						logger.Info.Printf("Ignored event: %s\n", ev)
					}
			}
		}
	}()
`

Checklist
=========
Please provide the following information:

 - [ ] confluent-kafka-go and librdkafka version (librdkafka1-0.11.0_confluent3.3.0-1.el6.x86_64):
 - [ ] Apache Kafka broker version: 0.9.1
 - [ ] Client configuration: `&kafka.ConfigMap{
			"bootstrap.servers": strings.Join(conf.Conf.Kafka.Brokers, ", "),
			"message.max.bytes": 1024 * 1024 * 40,
			"go.batch.producer": true,
		})`
 - [ ] Operating system: CentOS 6.5/Ubuntu 17.04
 - [ ] Provide client logs (with `"debug": ".."` as necessary)
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
Sorry for the slow response, dropped this somehow.

What I think is happening is that it eventually hangs on this line `producer.ProduceChannel() <- &kafka.Message` because the ProduceChannel is full, which is caused the batcher's produceBatch() call blocking, because the maximum number of messages in librdkafka  has been reached (`queue.buffering.max.messages` - this number accounts both for to-be-produced messages as well as delivery reports back to the application). To relieve the pressure of messages from librdkafka the application needs to read the Events channel for delivery reports.
So this is effectively a channel dead-lock:
TLDR: The ProduceChannel is full because the Events channel is not read from.

The workaround is to use non-blocking sending to ProduceChannel and if it fails (channel buffer full) you  read as many events from the Events channel (non-blockingly too) and try ProduceChannel again.

**Comment by enchantner** [comment]
But I actually read from Events simultaneously, see last case at the bottom. Or do you mean it just hangs on that line and another "case" inside "switch" never happens afterwards?

**Comment by enchantner** [comment]
All right, looks like things are better when I moved sending messages and receiving events to different goroutines. But now I've started getting lots of 
`INFO: 2018/01/11 04:17:57 main.go:103: Delivery failed: Local: Message timed out
INFO: 2018/01/11 04:17:57 main.go:103: Delivery failed: Local: Message timed out`

I've tried adding `"message.timeout.ms": 360000` to the producer but it didn't help much. What should I do?

**Comment by edenhill** [comment]
@enchantner librdkafka 0.11.0 had some issues with performance and I suggest you upgrade to librdkafka v0.11.3 (you don't need to rebuild the Go bindings (unless they are statically linked).

The default producer settings in librdkafka are aimed at low latency with linger.ms=1, but this gives poor throughput when trying to produce messages at a higher rate - if this is the case for you I suggest setting `linger.ms` to somewhere between 50-500ms depending on your delivery requirements.

**Comment by enchantner** [comment]
@edenhill thanks a lot! Looks like the last version available through Confluent yum repo is librdkafka-devel.x86_64 0:0.11.1_confluent4.0.0-1.el6 . I'll try to use this version, also will update `linger.ms` for producer and post an update here.

**Comment by enchantner** [comment]
@edenhill well, setting "linger.ms" to 100 didn't help, unfortunately. I'm using the comparison between the number of messages sent to channel and the number I received acknowledgements for. And the process sometimes passes flawlessly, sometimes hangs. Any other options?

**Comment by edenhill** [comment]
Try setting the `debug` config property to `msg` on the producer and let it run for a bit and then paste a typical excerpt, like 50-100 lines, here for further inspection.
